pkg/ovsdb/ovsdb.go

```diff
@@ -0,0 +1,311 @@
+// Create an internal port in OVS
+func (self *OvsBridgeDriver) CreatePort(infoMap map[string]string, vlanTag uint) error {
+	portUuidStr := infoMap["intfName"]
+	intfUuidStr := fmt.Sprintf("Intf%s", infoMap["intfName"])
+	portUuid := []libovsdb.UUID{{GoUUID: portUuidStr}}
+	intfUuid := []libovsdb.UUID{{GoUUID: intfUuidStr}}
+	opStr := "insert"
+	var err error = nil
+
+	// insert/delete a row in Interface table
+	intf := make(map[string]interface{})
+	intf["name"] = infoMap["intfName"]
+	//intf["type"] = ""
+
+	if infoMap["ovnPortName"] != "" {
+		oMap, err := libovsdb.NewOvsMap(map[string]string{"iface-id": infoMap["ovnPortName"]})
+		if err != nil {
+			return nil
+		}
+		intf["external_ids"] = oMap
+	}
+
+	// Add an entry in Interface table
+	intfOp := libovsdb.Operation{
+		Op:       opStr,
+		Table:    "Interface",
+		Row:      intf,
+		UUIDName: intfUuidStr,
+	}
+
+	// insert/delete a row in Port table
+	port := make(map[string]interface{})
+	port["name"] = infoMap["intfName"]
+	if vlanTag != 0 {
+		port["vlan_mode"] = "access"
+		port["tag"] = vlanTag
+	} else {
+		port["vlan_mode"] = "trunk"
+	}
+
+	port["interfaces"], err = libovsdb.NewOvsSet(intfUuid)
+	if err != nil {
+		return err
+	}
+
+	oMap, err := libovsdb.NewOvsMap(map[string]string{"contNetns": infoMap["contNetnsPath"], "contIface": infoMap["contIfaceName"]})
+	if err != nil {
+		return nil
+	}
+	port["external_ids"] = oMap
+
+	// Add an entry in Port table
+	portOp := libovsdb.Operation{
+		Op:       opStr,
+		Table:    "Port",
+		Row:      port,
+		UUIDName: portUuidStr,
+	}
+
+	// mutate the Ports column of the row in the Bridge table
+	mutateSet, _ := libovsdb.NewOvsSet(portUuid)
+	mutation := libovsdb.NewMutation("ports", opStr, mutateSet)
+	condition := libovsdb.NewCondition("name", "==", self.OvsBridgeName)
+	mutateOp := libovsdb.Operation{
+		Op:        "mutate",
+		Table:     "Bridge",
+		Mutations: []interface{}{mutation},
+		Where:     []interface{}{condition},
+	}
+
+	// Perform OVS transaction
+	operations := []libovsdb.Operation{intfOp, portOp, mutateOp}
+	return self.ovsdbTransact(operations)
+}
+
+// Delete a port from OVS
+func (self *OvsBridgeDriver) DeletePort(intfName string) error {
+	// lock the cache for read
+	self.lock.RLock()
+	defer self.lock.RUnlock()
+
+	condition := libovsdb.NewCondition("name", "==", intfName)
+	selectOp := libovsdb.Operation{
+		Op:    "select",
+		Table: "Port",
+		Where: []interface{}{condition},
+	}
+
+	transactionResult, err := self.ovsClient.Transact("Open_vSwitch", selectOp)
+	if err != nil {
+		return err
+	}
+
+	if len(transactionResult) != 1 {
+		return fmt.Errorf("unknow error")
+	}
+
+	operationResult := transactionResult[0]
+	if operationResult.Error != "" {
+		return fmt.Errorf("%s - %s", operationResult.Error, operationResult.Details)
+	}
+
+	if len(operationResult.Rows) != 1 {
+		return fmt.Errorf("failed to find port")
+	}
+
+	portUuidStr := operationResult.Rows[0]["_uuid"].([]interface{})
```

Comment:
Please comment on what are you doing here

Reply:
done